tests/concurrent_client_request_test.go

```diff
@@ -0,0 +1,141 @@
+package tests
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"sync"
+	"testing"
+	"time"
+
+	"google.golang.org/grpc"
+	"sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client"
+	"sigs.k8s.io/apiserver-network-proxy/pkg/agent/agentserver"
+	"sigs.k8s.io/apiserver-network-proxy/proto/agent"
+)
+
+type simpleServer struct {
+	receivedSecondReq chan struct{}
+}
+
+// ServeHTTP blocks the response to the request whose body is "1" until a
+// request whose body is "2" is handled.
+func (s *simpleServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	bytes, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		w.Write([]byte(err.Error()))
+	}
+	if string(bytes) == "2" {
+		close(s.receivedSecondReq)
+		w.Write([]byte("2"))
+	}
+	if string(bytes) == "1" {
+		<-s.receivedSecondReq
+		w.Write([]byte("1"))
+	}
+}
+
+// TODO: test http-connect as well.
+func getTestClient(front string, t *testing.T) *http.Client {
+	tunnel, err := client.CreateGrpcTunnel(front, grpc.WithInsecure())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return &http.Client{
+		Transport: &http.Transport{
+			Dial: tunnel.Dial,
+		},
+		Timeout: 2 * time.Second,
+	}
+}
+
+// singleTimeManager makes sure that a backend only serves one request.
+type singleTimeManager struct {
+	mu sync.Mutex
```

Comment:
Please change name to not match mu in DefaultBackendManager.

Reply:
Removed the embedding altogether.